rust/lance-encoding/src/encodings/logical/dictionary.rs

```diff
@@ -0,0 +1,192 @@
+            } else {
+                self.buffered_arrays.push(deep_copy_array(array.as_ref()))
+            }
+            None
+        }
+    }
+
+    pub fn flush(&mut self) -> Option<Vec<ArrayRef>> {
+        if self.buffered_arrays.is_empty() {
+            trace!(
+                "No final flush since no data at column {}",
+                self.column_index
+            );
+            None
+        } else {
+            trace!(
+                "Final flush of column {} which has {} bytes",
+                self.column_index,
+                self.current_bytes
+            );
+            self.current_bytes = 0;
+            Some(std::mem::take(&mut self.buffered_arrays))
+        }
+    }
+}
+
+pub struct DictionaryFieldEncoder {
+    accumulation_queue: AccumulationQueue,
+    indices_encoder: Box<dyn FieldEncoder>,
+    // items_encoder: Box<dyn FieldEncoder>,
+}
+
+impl DictionaryFieldEncoder {
+    pub fn new(cache_bytes: u64, max_items_per_column: u64, keep_original_array: bool, column_index: u32) -> Self {
+
+        // encoder for dictionary indices
+        let indices_encoder = Box::new(
+            PrimitiveFieldEncoder::try_new(
+                cache_bytes,
+                keep_original_array,
+                &DataType::UInt64,
+                column_index + 1,
+            )
+            .unwrap(),
+        );
+
+        // TODO: create encoder for dictionary items
+        // let items_encoder = 
+
+        // return overall dict encoder object
+        Self {
+            accumulation_queue: AccumulationQueue::new(
+                cache_bytes,
+                max_items_per_column,
+                column_index,
+                keep_original_array,
+            ),
+            indices_encoder,
+            // items_encoder,
+        }
+    }
+}
+
+impl FieldEncoder for DictionaryFieldEncoder {
+    fn maybe_encode(&mut self, array: ArrayRef) -> Result<Vec<EncodeTask>> {
+
+        println!("array: {:?}", array); // array of strings
+
+        let mut num_distinct_elements = 0;
+        // need to check cardinality of elements in this array
+        if let Some(string_array) = array.as_any().downcast_ref::<StringArray>() {
+            let mut distinct_elements = HashSet::new();
+        
+            for i in 0..string_array.len() {
+                if !string_array.is_null(i) {
+                    distinct_elements.insert(string_array.value(i));
+                }
+            }
+        
+            num_distinct_elements = distinct_elements.len();
+            println!("Number of distinct elements: {}", num_distinct_elements);
+        } 
+        // ideally need to extend to other datatypes as well. For now just testing on strings.
+        else { 
+            println!("The array is not a StringArray");
+        }
+
+        if num_distinct_elements < 100 {
+            // do dictionary encoding
+        }
+        else {
+            // use varbin encoder's maybe_encode function (in binary.rs)
+        }
```

Comment:
We probably want to do this check after we've accumulated the arrays (e.g. down below on line 175 or so)